proposals/gc/MVP-JS.md

```diff
@@ -1,538 +1,65 @@
-  for a given type definition `T`.
-* `WebAssembly.i31ref` : singleton reflecting [`i31ref`](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#value-types)
-  whose `[[Write]]` only accepts integral `Number` and `BigInt` values in the range [-2<sup>30</sup>, 2<sup>30</sup>-1].
-* `WebAssembly.rtt(T)` : constructor function reflecting [`rtt <typeidx>`](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#value-types),
-  for a given type definition `T`.
-
-Additional Value Type objects could be added in the future as WebAssembly's set
-of reference types evolve.
-
-
-## Other WebAssembly uses of Value Type objects
-
-In addition to the new struct and array typed objects, the existing
-`WebAssembly.Table` and `WebAssembly.Global` objects contain typed values that
-are read and written from JS and thus should be extended to accept any Value
-Type object. Reading and writing these values (via, e.g.,
-[`WebAssembly.Table.prototype.get`](http://webassembly.github.io/spec/js-api/index.html#dom-table-get)
-and [`set`](http://webassembly.github.io/spec/js-api/index.html#dom-table-set)
-or the [`WebAssembly.Global` `value` getter and setter](http://webassembly.github.io/spec/js-api/index.html#dom-global-value))
-invokes the Value Type object `[[Read]]` and `[[Write]]` methods symmetric
-to the Typed Object field access described above.
-
-
-## TODO
-
-* Should there be a `FunctionType` in the Typed Objects proposal, symmetric to
-  `StructType`/`ArrayType`, or should we have `WebAssembly.Function`
-  [as separately proposed](https://github.com/WebAssembly/js-types/blob/master/proposals/js-types/Overview.md#addition-of-webassemblyfunction)
-  (or both, with `WebAssembly.Function` subclassing `FunctionType`)?
-  Either way, the function type would be a [Type Definition object](#type-definition-objects),
-  whose signature was defined in terms of [Value Type objects](#value-type-objects).
-
-* How do [RTT values](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#runtime-types)
-  look when they flow out to JS? Are RTT values just first-class references to
-  [Typed Definition objects](#type-definition-objects)?
-
-* There is some hand-waving here around nullability / defaultability that needs
-  to be cleared up.
+The goal of the MVP JS API is to be as simple as possible, providing just enough
+expressivity to serve as the foundation for interop between JS and Wasm GC.
+Making it easy to access, manipulate, and extend Wasm GC objects from
+hand-written JS is left as future work. For now, require using the established
+pattern of exporting accessor functions to interact with Wasm data from JS.
+
+## Explanation
+
+WebAssembly GC objects are exposed to JS via opaque exotic object wrappers. The
+wrapper object internally holds a reference to the underlying WasmGC object to
+facilitate identity-preserving round-tripping from WebAssembly to JS and back.
```

Comment:
I believe this is a bug in the current formulation of the spec, since it says `If |internref| is of the form [=ref.struct=] |structaddr|, return the result of creating [=a new Exported GC Object=] from |structaddr| and "struct"`, i.e. it always creates a fresh wrapper object. I'll add a TODO here to fix this.

Reply:
While it's true that the cached lookup isn't done in that step, I think the creating "a new Exported GC Object" algorithm should handle this already. In particular, it looks up the address in an exported GC object cache and returns the cached wrapper object if it exists: https://github.com/WebAssembly/gc/blob/main/document/js-api/index.bs#L1262

Reply:
Aha, apologies for not reading that more closely. I'll update the doc to reflect that. Additionally, WDYT about removing "new" from that algorithm name, since it may in fact return a pre-existing wrapper object?

Reply:
No problem, thanks for updating. I don't have a strong opinion either way on the algorithm name but I think it'd be nice to keep it consistent; the reason why it has "new" currently is just to match the algorithm name for making function wrappers (https://github.com/WebAssembly/gc/blob/main/document/js-api/index.bs#L989). One option would be to change both to remove the "new".